pkg/services/authn/authnimpl/sync/oauth_token_sync.go

```diff
@@ -56,44 +63,44 @@ func (s *OAuthTokenSync) SyncOauthTokenHook(ctx context.Context, id *authn.Ident
 		return nil
 	}
 
-	ctxLogger := s.log.FromContext(ctx).New("userID", id.GetID())
+	userID, err := id.GetInternalID()
+	if err != nil {
+		s.log.FromContext(ctx).Error("Failed to refresh token. Invalid ID for identity", "type", id.GetIdentityType(), "err", err)
+		return nil
+	}
 
-	_, err, _ := s.singleflightGroup.Do(id.GetID(), func() (interface{}, error) {
+	ctxLogger := s.log.FromContext(ctx).New("userID", userID)
+
+	cacheKey := fmt.Sprintf("token-check-%s", id.GetID())
+	if _, ok := s.cache.Get(cacheKey); ok {
+		ctxLogger.Debug("Expiration check has been cached, no need to refresh")
+		return nil
+	}
+
+	_, err, _ = s.singleflightGroup.Do(id.GetID(), func() (interface{}, error) {
```

Comment:
We don't use singleflight much, but wouldn't hurt to namespace this too just in case
```suggestion
	_, err, _ = s.singleflightGroup.Do(cacheKey, func() (interface{}, error) {
```